code/game/objects/structures/displaycase.dm

```diff
@@ -396,64 +396,135 @@
 	src.icon = I
 	return
 
+/obj/structure/displaycase/forsale/ui_interact(mob/user, ui_key, datum/tgui/ui, force_open, datum/tgui/master_ui, datum/ui_state/state)
+	. = ..()
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "Vendatray", name, 300, 300, master_ui, state)
+		ui.open()
+
+/obj/structure/displaycase/forsale/ui_data(mob/user)
+	var/list/data = list()
+	var/register = FALSE
+	if(payments_acc)
+		register = TRUE
+		data["owner_name"] = payments_acc.account_holder
+	if(showpiece)
+		data["product_name"] = capitalize(showpiece.name)
+		var/base64 = icon2base64(icon(showpiece.icon, showpiece.icon_state))
+		data["product_icon"] = base64
+	data["registered"] = register
+	data["product_cost"] = sale_price
+	data["tray_open"] = open
+	return data
+
+/obj/structure/displaycase/forsale/ui_act(action, params)
+	if(..())
+		return
+	var/obj/item/card/id/potential_acc = usr.get_idcard(hand_first = TRUE)
+	switch(action)
+		if("Buy")
+			if(!showpiece)
+				to_chat(usr, "<span class='notice'>There's nothing for sale.</span>")
+				return TRUE
+			if(broken)
+				to_chat(usr, "<span class='notice'>[src] appears to be broken.</span>")
+				return TRUE
+			if(!payments_acc)
+				to_chat(usr, "<span class='notice'>[src] hasn't been registered yet.</span>")
+				return TRUE
+			var/confirm = alert(usr, "Purchase [showpiece] for [sale_price]?", "Purchase?", "Confirm", "Cancel")
+			if(confirm == "Cancel" || !usr.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
+				return TRUE
+			if(!potential_acc)
+				to_chat(usr, "<span class='notice'>No ID card detected.</span>")
+				return
+			var/datum/bank_account/account = potential_acc.registered_account
+			if(!account)
+				to_chat(usr, "<span class='notice'>[potential_acc] has no account registered!</span>")
+				return
+			if(!account.has_money(sale_price))
+				to_chat(usr, "<span class='notice'>You do not possess the funds to purchase this.</span>")
+				return TRUE
+			else
+				account.adjust_money(-sale_price)
+				if(payments_acc)
+					payments_acc.adjust_money(sale_price)
+				usr.put_in_hands(showpiece)

```

Comment:
what if they're using tk

Reply:
That's checked on 437 in the same instance, canUseTopic with NO_TK.